ariadne/cardano/src/Ariadne/MainTemplate.hs

```diff
@@ -157,23 +158,24 @@ initializeEverything MainSettings {..}
     initAction :: IO ()
     initAction = walletInitAction
 
+    postStartupAction :: IO ()
+    postStartupAction = walletAction
+
     serviceAction :: IO ()
     serviceAction =
-      raceWithUpdateCheckAction $
-      cardanoAction
+      raceWithUpdateCheckAction cardanoAction
 
     mainAction :: IO ()
     mainAction = do
       initAction
-
-      -- Spawn backend actions in async thread, then run ui action in the main thread
-      -- This is needed because some UI libraries (Qt) insist on livng in the main thread
-      withAsync serviceAction $ \serviceThread -> do
-        -- Make custom link to service thread for UI apps.
-        -- It is going to call msPutBackendErrorToUI and rethrow exception, if something goes wrong
-        linkUI serviceThread $ msPutBackendErrorToUI uiFace
-        uiAction
-
+      withAsync postStartupAction $ \postStartupThread -> do
+        linkUI postStartupThread $ msPutBackendErrorToUI uiFace
+        -- Spawn backend actions in async thread, then run ui action in the main thread
+        -- This is needed because some UI libraries (Qt) insist on livng in the main thread
+        withAsync serviceAction $ \serviceThread -> do
+          -- Make backend rethrow all exceptions in main thread if something goes wrong
```

Comment:
It looks like you edited this comment, but it's completely unrelated to this issue, AFAIU. Maybe it was by accident.

Reply:
Everything below the line with `withAsync` should be left intact (because it's irrelevant to this PR).

Reply:
Sorry. The change was reverted, but it seems that it returned during a rebase.